File: docs/DSA/src/py/diameter_of_binary_tree.py
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def _helper(self, node):
        if not node:
            return 0, 0
        left_height, left_diameter = self._helper(node.left)
        right_height, right_diameter = self._helper(node.right)
        height = 1+max(left_height, right_height)
        diameter = max(left_diameter, right_diameter, 1+left_height+right_height)
        return height, diameter

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
         _, diameter = self._helper(root)
         return diameter

File: docs/DSA/src/go/diameter_of_binary_tree.go
package main

func diameterOfBinaryTree(root *TreeNode) int {
	var _helper func(node *TreeNode) (int, int)

	_helper = func(node *TreeNode) (int, int) {
		if node == nil {
			return 0, 0
		}

		leftHeight, leftDiameter := _helper(node.Left)
		rightHeight, rightDiameter := _helper(node.Right)
		height := 1 + max(leftHeight, rightHeight)
		diameter := max(leftDiameter, rightDiameter, leftHeight+rightHeight)
		return height, diameter
	}

	_, diameter := _helper(root)
	return diameter
}
